Filter datastores for file volume creation in WCP

gohilankit opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
1. This PR is adding a new field `targetvSANFileShareClusters` to vSphere config secret. This field is optional and will be used only in WCP to filter datastores for file volume creation.  This field will be set by `wcpsvc` in the CSI secret when file volume feature is enabled on WCP.
The reason we're introducing this and not rely on existing `targetvSANFileShareDatastoreURLs` field is because `wcpsvc` doesn't want to add the logic fetch vSAN fileshare datastores in their layer. So the existing `targetvSANFileShareDatastoreURLs` field will be empty there.

2. This also adds a logic to periodically update the map of vSAN file service enabled cluster IDs to corresponding datastores, which will be used in file volume creation workflow.
For file volume creation in WCP, we currently take into account all clusters with Host.Config.Storage privilege and file services enabled. With introduction of this change, we will add another filter that only vSAN FS datastores corresponding to cluster IDs in the list will be used for file volume placement.

Since `targetvSANFileShareClusters` is not supposed to be used with vanilla set-up, file volume creation in vanilla will continue to use the same logic. It's just refactored a bit to use the map of cluster ID to datastore info objects.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Testing done**:

**WCP**: 
Enabled `csi-auth-check` and `file-volume` feature flags. The field `targetvSANFileShareClusters` was set in CSI secret by `wcpsvc`.
```
[Global]
insecure-flag = "true"
cluster-id = "domain-c9"
cnsregistervolumes-cleanup-intervalinmin = 720
cluster-distribution = "SupervisorCluster"
[VirtualCenter "sc-rdops-vm12-dhcp-124-153.eng.vmware.com"]
user = "workload_storage_management-1637b8a1-d93d-42af-bc4f-77318241a294@vsphere.local"
password = "%^E1dmI&vgzwU_m|F~b^"
datacenters = "datacenter-4"
port = "443"
targetvSANFileShareClusters = "domain-c9"         <<<<<<------------
```

With `csi-auth-check` enabled, the datastore maps generated periodically for file volumes are getting updated. Container init logs:
```
.
.
2021-04-09T00:53:51.327Z	DEBUG	common/authmanager.go:141	auth manager: refreshDatastoreMapsForFileVolumes is triggered	{"TraceId": "a5854276-495a-4089-9f71-4b90632d357b"}
2021-04-09T00:53:51.756Z	DEBUG	common/authmanager.go:390	Computing the map for vSAN FS enabled clusters to datastore URLS.	{"TraceId": "a5854276-495a-4089-9f71-4b90632d357b"}
2021-04-09T00:53:51.756Z	DEBUG	common/authmanager.go:433	Computing the clusters with vSAN file services enabled and Host.Config.Storage privileges	{"TraceId": "a5854276-495a-4089-9f71-4b90632d357b"}
2021-04-09T00:53:51.799Z	DEBUG	common/authmanager.go:473	auth manager: HasUserPrivilegeOnEntities returns [{{} ClusterComputeResource:domain-c21 [{{} Host.Config.Storage true}]} {{} ClusterComputeResource:domain-c9 [{{} Host.Config.Storage true}]}] when checking privileges [Host.Config.Storage] on entities [ClusterComputeResource:domain-c21 ClusterComputeResource:domain-c9] for user workload_storage_management-1637b8a1-d93d-42af-bc4f-77318241a294@vsphere.local	{"TraceId": "a5854276-495a-4089-9f71-4b90632d357b"}
2021-04-09T00:53:51.801Z	DEBUG	common/authmanager.go:489	Clusters with priv: Host.Config.Storage are : [ClusterComputeResource:domain-c21 @ /folder-WCP_DC/WCP_DC/host/edge-cluster ClusterComputeResource:domain-c9 @ /folder-WCP_DC/WCP_DC/host/compute-cluster]	{"TraceId": "a5854276-495a-4089-9f71-4b90632d357b"}
2021-04-09T00:53:51.861Z	DEBUG	common/authmanager.go:502	cluster: ClusterComputeResource:domain-c21 @ /folder-WCP_DC/WCP_DC/host/edge-cluster is a non-vSAN cluster. Skipping this cluster	{"TraceId": "a5854276-495a-4089-9f71-4b90632d357b"}
2021-04-09T00:53:51.892Z	DEBUG	common/authmanager.go:510	cluster: ClusterComputeResource:domain-c9 @ /folder-WCP_DC/WCP_DC/host/compute-cluster has vSAN file services enabled: true	{"TraceId": "a5854276-495a-4089-9f71-4b90632d357b"}
2021-04-09T00:53:51.925Z	DEBUG	common/authmanager.go:249	Adding vSAN datastore "ds:///vmfs/volumes/vsan:52ef5283816de19b-2ab5954ae5fb72d5/" to the list for FS enabled cluster "domain-c9"	{"TraceId": "a5854276-495a-4089-9f71-4b90632d357b"}
2021-04-09T00:53:51.931Z	DEBUG	common/authmanager.go:255	clusterToDsInfoListMap is map[domain-c9:[Datastore: Datastore:datastore-42, datastore URL: ds:///vmfs/volumes/vsan:52ef5283816de19b-2ab5954ae5fb72d5/]]	{"TraceId": "a5854276-495a-4089-9f71-4b90632d357b"}
2021-04-09T00:53:51.931Z	DEBUG	common/authmanager.go:148	auth manager: newFsEnabledClusterToDsMap is updated to map[domain-c9:[Datastore: Datastore:datastore-42, datastore URL: ds:///vmfs/volumes/vsan:52ef5283816de19b-2ab5954ae5fb72d5/]]	{"TraceId": "a5854276-495a-4089-9f71-4b90632d357b"}
.
.
```

When pvc was created with `ReadWriteMany` access mode, file volume did get created
Controller logs - The debug log captures that it's filtering the eligible datastores corresponding to the cluster ID in `targetvSANFileShareClusters`.

```
2021-04-09T00:55:47.465Z	INFO	wcp/controller.go:461	CreateVolume: called with args {Name:pvc-76d2346c-37d9-4b8a-b41d-efc119312e50 CapacityRange:required_bytes:1073741824  VolumeCapabilities:[mount:<> access_mode:<mode:MULTI_NODE_MULTI_WRITER > ] Parameters:map[storagePolicyID:ed1eb546-d70d-447b-a04b-16e693d8d7aa] ..
2021-04-09T00:55:47.476Z	DEBUG	wcp/controller.go:418	Adding datastore "ds:///vmfs/volumes/vsan:52ef5283816de19b-2ab5954ae5fb72d5/" to filtered datastores	{"TraceId": "f41e771e-871a-4dce-9433-988d8b321396"}
.
.
.
2021-04-09T00:55:59.114Z	INFO	volume/manager.go:307	CreateVolume: Volume created successfully. VolumeName: "pvc-76d2346c-37d9-4b8a-b41d-efc119312e50", opId: "7e27edeb", volumeID: "file:3468d791-0c64-4b89-a750-ee511574cfed"	{"TraceId": "f41e771e-871a-4dce-9433-988d8b321396"}
2021-04-09T00:55:59.116Z	DEBUG	volume/manager.go:308	CreateVolume volumeId "file:3468d791-0c64-4b89-a750-ee511574cfed" is placed on datastore "ds:///vmfs/volumes/vsan:52ef5283816de19b-2ab5954ae5fb72d5/"
```

**Vanilla**:
 Enabled `csi-auth-check` and `file-volume` feature flags. The datastore maps are updated periodically for file volumes. Container init logs:
 ```
 2021-04-09T01:24:07.749Z	INFO	common/authmanager.go:167	auth manager: ComputeFSEnabledClustersToDsMap enter
2021-04-09T01:24:07.749Z	DEBUG	common/authmanager.go:141	auth manager: refreshDatastoreMapsForFileVolumes is triggered	{"TraceId": "68c42a94-707e-4e84-a4eb-94d947f723f9"}
2021-04-09T01:24:07.814Z	DEBUG	common/authmanager.go:390	Computing the map for vSAN FS enabled clusters to datastore URLS.	{"TraceId": "68c42a94-707e-4e84-a4eb-94d947f723f9"}
2021-04-09T01:24:07.814Z	DEBUG	common/authmanager.go:433	Computing the clusters with vSAN file services enabled and Host.Config.Storage privileges	{"TraceId": "68c42a94-707e-4e84-a4eb-94d947f723f9"}
2021-04-09T01:24:07.842Z	DEBUG	common/authmanager.go:473	auth manager: HasUserPrivilegeOnEntities returns [{{} ClusterComputeResource:domain-c53 [{{} Host.Config.Storage true}]}] when checking privileges [Host.Config.Storage] on entities [ClusterComputeResource:domain-c53] for user Administrator@vsphere.local	{"TraceId": "68c42a94-707e-4e84-a4eb-94d947f723f9"}
2021-04-09T01:24:07.842Z	DEBUG	common/authmanager.go:489	Clusters with priv: Host.Config.Storage are : [ClusterComputeResource:domain-c53 @ /test-vpx-1616553346-30311-hostpool/host/vSAN-FVT-Cluster1616553475.519551]	{"TraceId": "68c42a94-707e-4e84-a4eb-94d947f723f9"}
2021-04-09T01:24:07.930Z	DEBUG	common/authmanager.go:510	cluster: ClusterComputeResource:domain-c53 @ /test-vpx-1616553346-30311-hostpool/host/vSAN-FVT-Cluster1616553475.519551 has vSAN file services enabled: true	{"TraceId": "68c42a94-707e-4e84-a4eb-94d947f723f9"}
2021-04-09T01:24:07.946Z	DEBUG	common/authmanager.go:249	Adding vSAN datastore "ds:///vmfs/volumes/vsan:5275b948f0481b91-81f8515679f6c850/" to the list for FS enabled cluster "domain-c53"	{"TraceId": "68c42a94-707e-4e84-a4eb-94d947f723f9"}
2021-04-09T01:24:07.947Z	DEBUG	common/authmanager.go:255	clusterToDsInfoListMap is map[domain-c53:[Datastore: Datastore:datastore-59, datastore URL: ds:///vmfs/volumes/vsan:5275b948f0481b91-81f8515679f6c850/]]	{"TraceId": "68c42a94-707e-4e84-a4eb-94d947f723f9"}
2021-04-09T01:24:07.947Z	DEBUG	common/authmanager.go:148	auth manager: newFsEnabledClusterToDsMap is updated to map[domain-c53:[Datastore: Datastore:datastore-59, datastore URL: ds:///vmfs/volumes/vsan:5275b948f0481b91-81f8515679f6c850/]]	{"TraceId": "68c42a94-707e-4e84-a4eb-94d947f723f9"}
```

When pvc was created with `ReadWriteMany` access mode, file volume did get created
Controller logs:

```
2021-04-09T01:31:03.464Z	INFO	vanilla/controller.go:625	CreateVolume: called with args {Name:pvc-5c12922a-0929-4379-aa49-2bbd2f907c29 CapacityRange:required_bytes:1073741824  VolumeCapabilities:[mount:<fs_type:"ext4" > access_mode:<mode:MULTI_NODE_MULTI_WRITER > ] Parameters:map[] Secrets:map[] VolumeContentSource:<nil> AccessibilityRequirements:<nil> XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}	{"TraceId": "5c6cf200-7279-427f-a8a7-0f0ef662f6a6"}
2021-04-09T01:31:03.464Z	DEBUG	vsphere/utils.go:385	Checking if vCenter version is of vsan 67u3 release	{"TraceId": "5c6cf200-7279-427f-a8a7-0f0ef662f6a6"}
.
.
2021-04-09T01:31:15.303Z	INFO	volume/manager.go:307	CreateVolume: Volume created successfully. VolumeName: "pvc-5c12922a-0929-4379-aa49-2bbd2f907c29", opId: "2a17ebb7", volumeID: "file:92880b13-ba85-480d-a086-5d1c653a888b"	{"TraceId": "5c6cf200-7279-427f-a8a7-0f0ef662f6a6"}
2021-04-09T01:31:15.303Z	DEBUG	volume/manager.go:308	CreateVolume volumeId "file:92880b13-ba85-480d-a086-5d1c653a888b" is placed on datastore "ds:///vmfs/volumes/vsan:5275b948f0481b91-81f8515679f6c850/"	{"TraceId": "5c6cf200-7279-427f-a8a7-0f0ef662f6a6"}
```



 
 


**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
Filter datastores for file volume creation in WCP
```

gohilankit commented:
jtest file-vanilla
jtest wcp

svcbot-qecnsdp commented:
Started vanilla file pipeline...

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
File vanilla build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 
S
Ran 26 of 165 Specs in 2695.003 seconds
SUCCESS! -- 26 Passed | 0 Failed | 0 Pending | 139 Skipped
PASS

Ginkgo ran 1 suite in 45m16.35231067s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/github-csi-file-vanilla/Results/240/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
``` 
WCP build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 
SSSSSSSSSSSSSSSSSSSSSSS
Ran 31 of 165 Specs in 12469.493 seconds
SUCCESS! -- 31 Passed | 0 Failed | 0 Pending | 134 Skipped
PASS

Ginkgo ran 1 suite in 3h40m26.816232124s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/552/vsphere-csi-driver`

```

divyenpatel commented on docs/book/driver-deployment/installation.md:
```diff
@@ -54,7 +54,6 @@ Here is an example vSphere configuration file for block volumes, with dummy valu
 $ cat /etc/kubernetes/csi-vsphere.conf
 [Global]
 cluster-id = "<cluster-id>"
-cluster-distribution = "<cluster-distribution>"
```
Can you add back documentation regarding cluster-distribution?

divyenpatel commented on docs/book/driver-deployment/installation.md:
```diff
@@ -93,8 +86,6 @@ Where the entries have the following meaning:
 - `datacenters` - list of all comma separated datacenter paths where kubernetes node VMs are present. When datacenter is located at the root, the name of datacenter is enough but when datacenter is placed in the folder, path needs to be specified as `folder/datacenter-name`.
 Please note since comma is used as a delimiter, the datacenter name itself must not contain a comma.
 
-**Note:** To deploy CSI driver for block volume in VMC environment, in the vSphere configuration file, need to specifiy cloudadmin user in `user` field and cloudadmin password in `password` field.
```
Can you add back support statement regarding VMC?

divyenpatel commented on pkg/common/cns-lib/vsphere/virtualcenter.go:
```diff
@@ -112,6 +112,8 @@ type VirtualCenterConfig struct {
 	// TargetDatastoreUrlsForFile represents URLs of file service enabled vSAN datastores in the virtual center.
 	TargetvSANFileShareDatastoreURLs []string
+	// TargetvSANFileShareClusters represents file service enabled vSAN clusters on which file volumes can be created.
+	TargetvSANFileShareClusters []string
```
`TargetvSANFileShareDatastoreURLs` already contains a list of vSAN datastore URLs on which file service is enabled. 
Why do we need one more field `TargetvSANFileShareClusters`? Having two fields for the same purpose will create confusion.

gohilankit replied:
So this new field will be set when file volume services are enabled by user in WCP. There's an API that accepts the list of cluster IDs. Now `wcpsvc` doesn't want the logic to fetch datastores corresponding to these clusters in their layer. They just agreed to pass the user input to CSI without any additional handling.
That's the reason we're introducing this new field. I have also updated the PR description for this.

divyenpatel replied:
So `TargetvSANFileShareClusters` should not be used in the vanilla setup? Can we add validation to disallow using this config for vanilla deployment?

gohilankit replied:
Yes it should not be used in vanilla setup. And we will not be adding info about this field in our fields.
Even if it's mentioned for vanilla, there will be no effect as we're not using it in vanilla. Should we still add the validation?

gohilankit commented:
jtest wcp

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 
SS
Ran 31 of 166 Specs in 11904.108 seconds
SUCCESS! -- 31 Passed | 0 Failed | 0 Pending | 135 Skipped
PASS

Ginkgo ran 1 suite in 3h20m49.602673505s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/566/vsphere-csi-driver`

```

BaluDontu commented on pkg/csi/service/common/authmanager.go:
```diff
@@ -284,7 +310,7 @@ func IsFileServiceEnabled(ctx context.Context, datastoreUrls []string, vc *vsphe
 		}
 	}
-	return dsToFSEnabledMapToReturn, nil
+	return dsToFSEnabledMapToReturn, clusterToFileServiceEnabledDsMap, nil
```
clusterToFileServiceEnabledDsMap naming is not correct. It is cluster which is FS enabled and not datastore. So 
`FsEnabledClusterToDsMap` sounds good.

here and every where.

gohilankit replied:
Makes sense. Will change it at all places.

gohilankit replied:
Done.

BaluDontu commented on pkg/csi/service/common/authmanager.go:
```diff
@@ -325,17 +351,20 @@ func getDatastoresWithBlockVolumePrivs(ctx context.Context, vc *vsphere.VirtualC
+// Returns
+// i. a map of vsan datastores to file service status (enabled/disabled).
+// ii. a map of cluster ID to file service enabled vsan datastore URLs.
+func createFSEnabledDsMapsForVC(ctx context.Context, vc *vsphere.VirtualCenter) (map[string]bool, map[string][]string, error) {
```
function name doesn't look good based on previous comment.

gohilankit replied:
But now we're returning multiple maps, and this takes into account all datastores for the VC.
What do you think can be an appropriate name?

BaluDontu replied:
`createFSEnabledDsMapsForVC` returning both maps in this function doesn't make sense.

Can you split it into 2 different functions. what we need here is clusters which have FS enabled and datastores within those clusters.

First function is `getFSEnabledClusters`
second function is `getDatastores([]clusters)`

createFSEnabledDsMapsForVC doesn't give the right sense.

BaluDontu commented:
@gohilankit Overall can you rename the variables and functions to align with the functionality.

gohilankit commented on pkg/csi/service/common/authmanager.go:
```diff
@@ -160,13 +181,13 @@ func ComputeDatastoreMapForBlockVolumes(authManager *AuthManager, authCheckInter
-// ComputeDatastoreMapForFileVolumes refreshes DatastoreMapForFileVolumes periodically
-func ComputeDatastoreMapForFileVolumes(authManager *AuthManager, authCheckInterval int) {
+// ComputeDatastoreMapsForFileVolumes refreshes DatastoreMapForFileVolumes & clusterToFSDatastoreURLMap periodically
+func ComputeDatastoreMapsForFileVolumes(authManager *AuthManager, authCheckInterval int) {
```
Since we're now computing two maps, making it plural to `Maps` seems logical. Also both the maps are related to datastores - one is `dsURL` to `dsInfo` map, the other one is cluster to list of corresponding datastores.

 > ComputeDatastoreMapForFileVolumes is better.
 
This doesn't imply we're computing multiple maps. Or we can just call it `ComputeMapsForFileVolumes`. Let me know if that makes sense.

gohilankit commented on pkg/csi/service/common/authmanager.go:
```diff
@@ -207,30 +228,32 @@ func GenerateDatastoreMapForBlockVolumes(ctx context.Context, vc *vsphere.Virtua
+// GenerateDatastoreMapsForFileVolumes scans all datastores in Vcenter and do privilege check on those datastoes
+// It will return datastores which has the privileges for creating file volume, as well as a mapping of VC clusters
+// against those datastores.
+func GenerateDatastoreMapsForFileVolumes(ctx context.Context, vc *vsphere.VirtualCenter) (map[string]*cnsvsphere.DatastoreInfo, map[string][]string, error) {
```
As clarified above, this meant to imply that we're generating multiple maps related to datastores for file volumes.

gohilankit commented on pkg/csi/service/vanilla/controller.go:
```diff
@@ -107,7 +107,7 @@ func (c *controller) Init(config *cnsconfig.Config, version string) error {
 	if !isAuthCheckFSSEnabled && len(c.manager.VcenterConfig.TargetvSANFileShareDatastoreURLs) > 0 {
 		// Check if file service is enabled on datastore present in targetvSANFileShareDatastoreURLs.
-		dsToFileServiceEnabledMap, err := common.IsFileServiceEnabled(ctx, c.manager.VcenterConfig.TargetvSANFileShareDatastoreURLs, vc)
+		dsToFileServiceEnabledMap, _, err := common.CreateFSEnabledDsMaps(ctx, c.manager.VcenterConfig.TargetvSANFileShareDatastoreURLs, vc)
```
Same as above. It's returning two maps now. Also I feel the earlier name `IsFileServiceEnabled` doesn't give a connotation that it returns a map.
Can you think of a better name?

BaluDontu commented:
Overall logic looks good. Just have a bias on the function names used. I will let the next reviewer take a look at them if its mandatory to change or not.

/approve.

gohilankit commented:
@BaluDontu I split it into multiple functions and refactored the code a bit for vanilla file volume creation also. This simplifies things. Can you take another stab at it?

To summarize, we no longer create a map of datastore url to info objects. We instead directly create a map of vSAN file service enabled cluster Ids to datastore info objects in `GenerateFSEnabledClustersToDsMap` which will be invoked periodically. This map will be used for file volume creation in both WCP and vanilla flavors.

gohilankit commented:
jtest wcp
jtest file-vanilla

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
Started vanilla file pipeline...

svcbot-qecnsdp commented:
``` 
File vanilla build status: FAILURE 
Stage before exit: testbed-deploy 

```

BaluDontu commented on pkg/csi/service/common/authmanager.go:
```diff
@@ -221,27 +225,35 @@ func GenerateDatastoreMapForFileVolumes(ctx context.Context, vc *vsphere.Virtual
-				dsURLToInfoMap[dsURL] = dsInfo
+	// Create a map of cluster to dsInfo objects. These objects are used while calling CNS
+	// API to create file volumes
+	for cluster, dsURLs := range fsEnabledClusterToDsURLsMap {
```
can you use clusterMoId instead of cluster.

BaluDontu commented on pkg/csi/service/common/authmanager.go:
```diff
@@ -326,16 +339,103 @@ func getDatastoresWithBlockVolumePrivs(ctx context.Context, vc *vsphere.VirtualC
 }
 
 // Creates a map of vsan datastores to file service status (enabled/disabled).
```
this is not correct right as this only contains vsan datastores which have file service enabled. so there will be no disabled status at all.

BaluDontu commented on pkg/csi/service/common/authmanager.go:
```diff
@@ -326,16 +339,103 @@ func getDatastoresWithBlockVolumePrivs(ctx context.Context, vc *vsphere.VirtualC
+	for _, cluster := range clustersWithPrivAndFS {
+		var dsList []vim25types.ManagedObjectReference
+		var dsMoList []mo.Datastore
+		datastores, err := cluster.Datastores(ctx)
```
retreiving datastore managed object from properties is repeated in `getDsToFileServiceEnabledMap` and `getFSEnabledClusterToDsURLsMap`. Can you create a separate function to retrieve datastore managed objects so that both these functions can use it.

This way that piece of code is not repeated.

gohilankit replied:
Makes sense. Done.

BaluDontu reviewed (commented):
Overall looks great with regards to readability of the code. Thanks for addressing it @gohilankit 

Just some minor comments and slight changes in refactoring.

gohilankit commented:
jtest file-vanilla
jtest wcp

svcbot-qecnsdp commented:
Started vanilla file pipeline...

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
File vanilla build status: FAILURE 
Stage before exit: testbed-deploy 

```

BaluDontu commented:
Code looks good for me. Approved from my side based on new changes.
/approve

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

BaluDontu commented:
@gohilankit Since this is touching the core component of the code, make sure to have all the pipeline succeed before checkin.

gohilankit commented:
jtest file-vanilla
jtest wcp

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
Started vanilla file pipeline...

svcbot-qecnsdp commented:
``` 
File vanilla build status: FAILURE 
Stage before exit: testbed-deploy 

```

svcbot-qecnsdp commented:
``` 
WCP build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 
SS
Ran 31 of 167 Specs in 6033.586 seconds
SUCCESS! -- 31 Passed | 0 Failed | 0 Pending | 136 Skipped
PASS

Ginkgo ran 1 suite in 1h42m16.945866082s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/580/vsphere-csi-driver`

```

gohilankit commented:
jtest file-vanilla

svcbot-qecnsdp commented:
Started vanilla file pipeline...

svcbot-qecnsdp commented:
``` 
File vanilla build status: FAILURE 
Stage before exit: testbed-deploy 

```

chethanv28 commented on pkg/csi/service/common/authmanager.go:
```diff
@@ -221,27 +225,35 @@ func GenerateDatastoreMapForFileVolumes(ctx context.Context, vc *vsphere.Virtual
+	fsEnabledClusterToDsURLsMap, err := getFSEnabledClusterToDsURLsMap(ctx, vc)
 	if err != nil {
-		log.Errorf("failed to get if file service is enabled on vsan datastores with error %+v", err)
+		log.Errorf("failed to get if file service enabled clusters map with error %+v", err)
```
log.Errorf("failed to get if file service enabled clusters map with error %+v", err) ==> log.Errorf("failed to get file service enabled clusters map with error %+v", err)

chethanv28 commented on pkg/csi/service/common/authmanager.go:
```diff
@@ -325,17 +338,80 @@ func getDatastoresWithBlockVolumePrivs(ctx context.Context, vc *vsphere.VirtualC
+			return nil, err
+		}
+		for _, dsMo := range dsMoList {
+			if dsMo.Summary.Type == VsanDatastoreType {
```
Can you confirm if we need a not nil check on `dsMo.Summary` ?

gohilankit replied:
No it won't be needed because we're fetching `summary` using property collector in `getDatastoreMOsFromCluster` method that's called to fetch `dsMoList`.
I have tested this for non-vsan datastores also and it doesn't panic.

chethanv28 commented on pkg/csi/service/common/authmanager.go:
```diff
@@ -325,17 +338,80 @@ func getDatastoresWithBlockVolumePrivs(ctx context.Context, vc *vsphere.VirtualC
+	log.Debugf("Computing the clusters with vSAN file services enabled and Host.Config.Storage privileges")
+	datacenters, err := vc.ListDatacenters(ctx)
+	if err != nil {
+		log.Errorf("failed to find datacenters from VC: %+v, Error: %+v", vc.Config.Host, err)
```
`vc.Config.Host` is %q ?

chethanv28 reviewed (commented):
few minor comments

chethanv28 commented:
/ok-to-test

gohilankit commented:
jtest file-vanilla

chethanv28 reviewed (approved):
looks good to me

svcbot-qecnsdp commented:
Started vanilla file pipeline...

svcbot-qecnsdp commented:
Started vanilla file pipeline...

svcbot-qecnsdp commented:
Started vanilla file pipeline...

svcbot-qecnsdp commented:
``` 
File vanilla build status: FAILURE 
Stage before exit: Parallel `images` block 

```

svcbot-qecnsdp commented:
``` 
File vanilla build status: FAILURE 
Stage before exit: Parallel `images` block 

```

svcbot-qecnsdp commented:
Started vanilla file pipeline...

svcbot-qecnsdp commented:
``` 
File vanilla build status: FAILURE 
Stage before exit: testbed-deploy 

```

svcbot-qecnsdp commented:
Started vanilla file pipeline...

gohilankit commented:
jtest file-vanilla

svcbot-qecnsdp commented:
Started vanilla file pipeline...

svcbot-qecnsdp commented:
``` 
File vanilla build status: FAILURE 
Stage before exit: testbed-deploy 

```

svcbot-qecnsdp commented:
``` 
File vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
Ran 27 of 167 Specs in 2363.324 seconds
FAIL! -- 26 Passed | 1 Failed | 0 Pending | 140 Skipped
--- FAIL: TestE2E (2363.39s)
FAIL

Ginkgo ran 1 suite in 41m32.143202738s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-file-vanilla/Results/270/vsphere-csi-driver`

```

gohilankit commented:
E2E tests status
WCP pipeline passed - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/760#issuecomment-817626312
File vanilla pipeline - 
```
Ran 27 of 167 Specs in 2363.324 seconds
FAIL! -- 26 Passed | 1 Failed | 0 Pending | 140 Skipped
```

The 1 failure is about CSI telemetry changes, and not related to this change.
`13:37:38  [Fail] [csi-block-vanilla] [csi-file-vanilla] CNS-CSI Cluster Distribution Telemetry [It] Verify dynamic provisioning of pvc has cluster-distribution value updated `

@BaluDontu @chethanv28 - Can you help merge this PR?

chethanv28 reviewed (commented):
Code changes and testing looks good to me
/approve
/lgtm
